src/resolve/dnstls-gnutls.c

```diff
@@ -0,0 +1,170 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#if !ENABLE_DNS_OVER_TLS || !HAVE_GNUTLS
+#error This source file requires DNS-over-TLS to be enabled and GnuTLS to be available.
+#endif
+
+#include "dnstls.h"
+#include "resolved-dns-stream.h"
+
+#include <gnutls/gnutls.h>
+#include <gnutls/socket.h>
+
+struct DnsTlsServerData {
+        gnutls_certificate_credentials_t tls_cert_cred;
+        gnutls_datum_t tls_session_data;
+};
+
```

Comment:
Beside a DnsTlsServerData struct a DnsTlsStreamData struct would be useful, so tls_bye and tls_handshake can be moved out of DnsStream